test/apps/iodemo/ucx_wrapper.cc

```diff
@@ -508,14 +512,73 @@ void UcxContext::set_am_handler(ucp_am_recv_callback_t cb, void *arg)
     ucp_worker_set_am_recv_handler(_worker, &param);
 }
 
+UcxEstablishCallback::UcxEstablishCallback(UcxConnection &connection) :
+    _connection(&connection),
+    _send_id_req(UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM)),
+    _recv_id_req(UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM)),
+    _comp_cnt(std::numeric_limits<unsigned>::max()),
+    _status(UCS_ERR_INVALID_PARAM)
+{
+    _connection->_established_cb = this;
+}
+
+UcxEstablishCallback::~UcxEstablishCallback()
+{
+    assert(_comp_cnt == 0);
+    _connection->_established_cb = NULL;
+}
+
+void UcxEstablishCallback::init(ucs_status_ptr_t send_id_req,
+                                ucs_status_ptr_t recv_id_req)
+{
+    assert(_send_id_req == UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM));
+    assert(_recv_id_req == UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM));
+    assert(_comp_cnt == std::numeric_limits<unsigned>::max());
+
+    _send_id_req = send_id_req;
+    _recv_id_req = recv_id_req;
+    _comp_cnt    = 2;
+    _status      = UCS_INPROGRESS;
+}
+
+void UcxEstablishCallback::operator()(ucs_status_t status)
+{
+    --_comp_cnt;
+    if (UCS_PTR_IS_PTR(_send_id_req) &&
+        ucp_request_check_status(_send_id_req) != UCS_INPROGRESS) {
+        _send_id_req = NULL;
+    }
+
+    if (UCS_PTR_IS_PTR(_recv_id_req) &&
+        ucp_request_check_status(_recv_id_req) != UCS_INPROGRESS) {
+        _recv_id_req = NULL;
+    }
+
+    if (_comp_cnt > 0) {
+        if (status != UCS_OK) {
+            _status = status;
+        }
+        return;
+    } else if (!UCS_STATUS_IS_ERR(_status)) {
+        _status = status;
+    }
+
+    if (_status == UCS_OK) {
```

Comment:
in this scope `_status` is a status of the Callback not a connections, need to complete it first then manage connection state. Otherwise we should manage resources from 2 places.